STM32 Bootloader support with retained ram boot mode infrastructure

petejohanson opened the pull request:
This is the initial work to support entering the STM32 built-in bootloader when the boot mode is set with the retained mem APIs. The PR assumes that a given platform will have the ROM bootloader addresss specified in the DTS using the values from AN2606 as a reference for the given target.

The *code* in question is working great on f072, but probably imperfect; I'd love some testing input to try to get it more bulletproof on other STM32 targets.

My targets stm32f072 devices work *perfectly* with this change. My stm32f401 and stm32f411 devices *do not*, they reset and attempt to jump to the bootloader but fail to enumerate. Sadly, neither of my f4 devices have SWD pins accessible, so I've not been able to further diagnose for those targets. Given that I'm *only* including the DTS changes to enable this for f072, pending further work.

petejohanson committed:
soc: arm: stm32: Add ROM bootloader support

Add jump to ROM bootloader for STM32 when bootmode is set via
retention boot mode feature.

Signed-off-by: Peter Johanson <peter@peterjohanson.com>

petejohanson committed:
arm: stm32: Add DTS describing ROM bootloader for f0

Add the ROM bootloader devicetree details for a stm32f072 SoC.

Signed-off-by: Peter Johanson <peter@peterjohanson.com>

petejohanson committed:
dts: arm: st: Add STM32WB bootloader information

Add the STM32 bootloader details to the relevant STM32WB .dtsi file.

Signed-off-by: Peter Johanson <peter@peterjohanson.com>

petejohanson committed:
boards: st: Add boot mode retention support to Nucleo WB55RG board

Setup a section of retained memory for boot mode retention on the Nucleo
WB55RG board.

Signed-off-by: Peter Johanson <peter@peterjohanson.com>

petejohanson committed:
arm: stm32: Add DTS describing ROM bootloader for f411

Add the ROM bootloader devicetree details for a stm32f411 SoC.

Signed-off-by: Peter Johanson <peter@peterjohanson.com>

henrikbrixandersen commented:
> Notes: Currently, I'm forcing off the RETENTION and RETAINED_MEM mutex Kconfig symbols to test this, but based on @henrikbrixandersen advice I think the more correct way is to update those existing portions of the code to skip the mutex logic if `k_is_in_isr() || k_is_pre_kernel()`. I'm happy to include that testing here, or do that in a separate PR.

Actually, I only suggested to skip the mutex locking in the retention subsystem when called in pre-kernel mode. I don't think it will always be safe to call these APIs in interrupt context due to the nature of the various backends.

I've opened a simple PR for this, which I actually prepared at ZDS, but never got around to pushing: https://github.com/zephyrproject-rtos/zephyr/pull/97012

henrikbrixandersen commented on soc/st/stm32/common/stm32_bootloader.c:
```diff
@@ -0,0 +1,57 @@
+	return 0;
+}
+
+SYS_INIT(bootloader_check_boot_init, EARLY, 0);
```
Won't you run into initialisation order issues here? The retention subsystem (and its backend drivers) are in post-kernel.

petejohanson replied:
So.... Yes, this is challenging. The stm32 bootloader, based on my research, can be persnickety if you try to jump to it with various things already init'd, like clocks, etc.

To be as robust as possible, I wanted to do this super early after a reset. The side effect is consuming device(s) that aren't technically in init'd yet, but we get *lucky* that the RAM based retained memory driver I'm using for my testing manages to work in this scenario. 

Some options:
* Move this stm32 bootloader code later, and enhance the code to properly reset/unload things before jumping.
* Move retention/retained memory bits earlier.
* Leave as is and it happens to work.

There's no real clear winner here, afaict

henrikbrixandersen replied:
> Some options:
> 
> * Move this stm32 bootloader code later, and enhance the code to properly reset/unload things before jumping.

That's certainly an option, but it might be complicated to deinitialize  everything.

> * Move retention/retained memory bits earlier.

I think this would be a good solution, but I doubt we can move it all the way to the `EARLY` initialisation level. @nordicjm: Any thoughts on this one?

> * Leave as is and it happens to work.

This is not an option. We cannot disregard the whole device model and initialisation order and rely on something that just happens to work.

nordicjm replied:
It should not be moved to earlier than `POST_KERNEL` to allow for drivers to init, e.g. it might need i2c. There already is driver deinit functionality and some drivers have it so makes sense to use that

petejohanson replied:
So far, the *cleanest*, but perhaps annoying to get right for all platforms, would be to move this init hook to a later level/priority that is correct from a sequence standpoint, and see what de-init is really needed. I will experiment with this.

If *device PM* is enabled, we could potentially leverage that to shut down any devices that are already loaded? Lacking that, I'm not sure how best to do this generically without just making the STM32 LL calls ourselves like `LL_RCC_DeInit`, etc.

petejohanson committed:
retention: Skip mutex usage when in pre-kernel

For early boot mode checking during initalization, skip mutex lock/unlock
if in pre-kernel state for the retention and retained mem zephyr ram
driver.

Signed-off-by: Peter Johanson <peter@peterjohanson.com>

nordicjm commented on drivers/retained_mem/Kconfig:
```diff
@@ -15,6 +15,11 @@ config RETAINED_MEM_INIT_PRIORITY
 	help
 	  Retained memory devices initialization priority,
 
+config RETAINED_MEM_INIT_LEVEL_PRE_KERNEL_1
+	bool "Retained memory devices init during PRE_KERNEL_1"
+	help
+	  Retained memory devices initialization during PRE_KERNEL_1 level
```
I really do not agree with this, this just makes blind assumptions about drivers which is impossible depending upon the driver

erwango replied:
If it depends on the driver, that is just that some driver may not be compatible, but I don't see it as a major issue.

petejohanson replied:
We could add a new Kconfig symbol, `RETAINED_MEM_SUPPORTS_PRE_KERNEL_INIT` that any compatible drivers could select, and have this symbol depend on that. Likewise, the retention config would also depend on *this* symbol, so the entire "do init early" could only be enabled with a driver that allows it.

Would that address your concerns?

nordicjm reviewed (changes requested):
Do not agree with changing init level of retention/retained memory

asm5878 commented:
@rjayles Could you please have a look at this PR?

petejohanson commented:
Hoping @erwango can revisit this at some point and give their thoughts.

etienne-lms commented on boards/st/nucleo_wb55rg/nucleo_wb55rg.dts:
```diff
@@ -24,6 +30,27 @@
+			};
+		};
+	};
+
```
Could you remove the extra empty line?

etienne-lms commented on soc/st/stm32/common/stm32_bootloader.c:
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2020 Google LLC.
```
Applicable copyright?

JarmouniA reviewed (commented):
https://github.com/zephyrproject-rtos/zephyr/pull/84272#discussion_r2404114280 is still not addressed.
